src/cuda/exec_dist.f90

```diff
@@ -0,0 +1,188 @@
+
+      integer :: n_block
+      integer :: ierr, tag = 1234
+      integer :: srerr(4), mpireq(4)
+
+      n_block = blocks%x
+
+      call der_univ_dist<<<blocks, threads>>>( &
+         du, du_send_s, du_send_e, u, u_recv_s, u_recv_e, &
+         tdsops%coeffs_s_dev, tdsops%coeffs_e_dev, tdsops%coeffs_dev, &
+         tdsops%n, tdsops%dist_fw_dev, tdsops%dist_bw_dev, tdsops%dist_af_dev &
+      )
+
+      ! halo exchange for 2x2 systems
+      if (nproc == 1) then
+         du_recv_s = du_send_e
+         du_recv_e = du_send_s
+      else
+         ! MPI send/recv for multi-rank simulations
+         call MPI_Isend(du_send_s, SZ*1*n_block, &
+                        MPI_DOUBLE_PRECISION, pprev, tag, MPI_COMM_WORLD, &
+                        mpireq(1), srerr(1))
+         call MPI_Irecv(du_recv_e, SZ*1*n_block, &
+                        MPI_DOUBLE_PRECISION, pnext, tag, MPI_COMM_WORLD, &
+                        mpireq(2), srerr(2))
+         call MPI_Isend(du_send_e, SZ*1*n_block, &
+                        MPI_DOUBLE_PRECISION, pnext, tag, MPI_COMM_WORLD, &
+                        mpireq(3), srerr(3))
+         call MPI_Irecv(du_recv_s, SZ*1*n_block, &
+                        MPI_DOUBLE_PRECISION, pprev, tag, MPI_COMM_WORLD, &
+                        mpireq(4), srerr(4))
+
+         call MPI_Waitall(4, mpireq, MPI_STATUSES_IGNORE, ierr)
+      end if
+
+      call der_univ_subs<<<blocks, threads>>>( &
+         du, du_recv_s, du_recv_e, &
+         tdsops%n, tdsops%dist_sa_dev, tdsops%dist_sc_dev &
+      )
+
+   end subroutine exec_dist_tds_compact
+
+   subroutine exec_dist_transeq_3fused( &
+      r_u, u, u_recv_s, u_recv_e, v, v_recv_s, v_recv_e, &
+      du, dud, d2u, &
+      du_send_s, du_send_e, du_recv_s, du_recv_e, &
+      dud_send_s, dud_send_e, dud_recv_s, dud_recv_e, &
+      d2u_send_s, d2u_send_e, d2u_recv_s, d2u_recv_e, &
+      der1st, der2nd, nu, nproc, pprev, pnext, blocks, threads &
+   )
+      implicit none
+
+      ! r_u = -1/2*(v*d1(u) + d1(u*v)) + nu*d2(u)
+      real(dp), device, dimension(:, :, :), intent(out) :: r_u
+      real(dp), device, dimension(:, :, :), intent(in) :: u, u_recv_s, u_recv_e
+      real(dp), device, dimension(:, :, :), intent(in) :: v, v_recv_s, v_recv_e
+
+      ! The ones below are intent(out) just so that we can write data in them,
+      ! not because we actually need the data they store later where this
+      ! subroutine is called. We absolutely don't care the data they pass back
+      real(dp), device, dimension(:, :, :), intent(out) :: du, dud, d2u
+      real(dp), device, dimension(:, :, :), intent(out) :: &
+         du_send_s, du_send_e, du_recv_s, du_recv_e, &
+         dud_send_s, dud_send_e, dud_recv_s, dud_recv_e, &
+         d2u_send_s, d2u_send_e, d2u_recv_s, d2u_recv_e
+
+      type(cuda_tdsops_t), intent(in) :: der1st, der2nd
+      real(dp), intent(in) :: nu
+      integer, intent(in) :: nproc, pprev, pnext
+      type(dim3), intent(in) :: blocks, threads
```

Comment:
What do you think about the argument ordering here?

Reply:
Some may argue that those temporary storages are relevant only in this subroutine, so maybe should be allocated in there, but we know that allocating temporary arrays each time a function is called is bad for performance.

Alternative would be wrapping this inside a class so that temporary data storages are a member of the class. But then we allocate temporaries for all the instances we create from this class, and this also results in unnecessary memory usage.

Reply:
So my point is grouping `r_u` and `du, dud, d2, ...` together and listing them next to each other just because they happen to have the same attribute `intent(out)` is wrong, because the semantic here is different for these two groups, and we need to figure out a way to communicate this for making it easy to understand.
So maybe the guideline for us could be
```
intent(out) [the ones we actually care about]
intent(in) [the corresponding fields we pass to get the result, and their auxilary components]
facilitators? [temporary storages that happen to have intent(out) attribute to please the compiler]
The rest of the inputs, scalars, coefficient arrays that we need to carry out the operation etc.
```

Any opinions? Would you prefer these `facilitators` between the actual `intent(out)`s and the corresponding `intent(in)`s? Is it a bad idea to seperate some of the `intent(in)`s from the rest of them like the scalar ones? What is the most natural ordering would be in this subroutine in your opinion?

Reply:
let's discuss about this on Monday :-)

Reply:
great, thanks